src/Type/Php/RegexArrayShapeMatcher.php

```diff
@@ -0,0 +1,233 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use Hoa\Compiler\Exception\Exception;
+use Hoa\Compiler\Llk\Llk;
+use Hoa\Compiler\Llk\Parser;
+use Hoa\Compiler\Llk\TreeNode;
+use Hoa\File\Read;
+use Nette\Utils\RegexpException;
+use Nette\Utils\Strings;
+use PHPStan\Php\PhpVersion;
+use PHPStan\TrinaryLogic;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function array_key_last;
+use function array_keys;
+use function count;
+use function in_array;
+use function is_int;
+use function is_string;
+use function str_contains;
+use const PREG_OFFSET_CAPTURE;
+use const PREG_UNMATCHED_AS_NULL;
+
+/**
+ * @api
+ */
+final class RegexArrayShapeMatcher
+{
+
+	private static ?Parser $parser = null;
+
+	public function __construct(
+		private PhpVersion $phpVersion,
+	)
+	{
+	}
+
+	public function matchType(Type $patternType, ?Type $flagsType, TrinaryLogic $wasMatched): ?Type
+	{
+		if (
+			!$this->phpVersion->returnsPregUnmatchedCapturingGroups()
+		) {
+			return null;
```

Comment:
We can't return anything useful on 7.2 and 7.3 at all?

Reply:
I have invested so much already in this PR, that I had not yet looked into 7.2/7.3 specifics. I am happy getting it running for the more modern stuff for now :)

Reply:
the array-shape matcher magic depends on named capturing groups.
in 7.2/7.3 these are not supported, so we can't provide a better array-shape for these php versions.

Reply:
just added emtpy-array negative context inference, which also works for php 7